Add 2nu ECEC generator

gonzaponte opened the pull request:
This PR is a resurrection of #26, which was closed due to repository configuration issues.

This PR adds a generator for 2-neutrino double electron capture.
Example macros have also been added for reference and testing.

gonzaponte committed:
Add ECECGenerator

gonzaponte committed:
Add ECEC example macros

paolafer commented on source/generators/ECECGenerator.cc:
```diff
@@ -0,0 +1,158 @@
+// ----------------------------------------------------------------------------
+// nexus | ECECGenerator.h
```
typo: `.h` --> `.cc`

paolafer commented on macros/NEW_2nuECEC.config.mac:
```diff
@@ -0,0 +1,23 @@
+### --------------------------------------------------------
+###  This init macro simulates bb0nu decays
```
Update this brief description both in the `init` and in the `config` file.

paolafer commented on macros/NEW_2nuECEC.config.mac:
```diff
@@ -0,0 +1,23 @@
+### --------------------------------------------------------
+###  This init macro simulates 2-neutrino ECEC decays
```
I know I'm a bit of a pain, but could you use the same format as the other example macros :-)?

gonzaponte replied:
You can't beat me at that ;). Do you mean like the one I just pushed?

paolafer commented on macros/NEW_2nuECEC.config.mac:
```diff
@@ -0,0 +1,27 @@
+## ----------------------------------------------------------------------------
+## nexus | NEW_2nuECEC.init.mac
```
Perfect! Just a typo: `init` --> `config` here.

gonzaponte replied:
Classic copy-paste mistake :)

paolafer commented on source/generators/ECECGenerator.cc:
```diff
@@ -0,0 +1,158 @@
+}
+
+
+G4AtomicShellEnumerator ECECGenerator::GetShellID(G4String s)
```
I get a compiler warning because of the variable name: `warning: declaration shadows a variable in namespace 'CLHEP'`. Could you use another name instead of `s`? Anything that is not related to units of measure.

paolafer commented on source/generators/ECECGenerator.cc:
```diff
@@ -0,0 +1,158 @@
+  shell_name_("K"), // K, L1, L2, L3, M1, ..., M5
+  region_("ACTIVE"),
+  msg_ (nullptr),
+
```
Remove blank line.

paolafer commented on source/generators/ECECGenerator.cc:
```diff
@@ -0,0 +1,157 @@
+  atom_ = new G4UAtomicDeexcitation();
+  atom_->SetFluo (true);
+  atom_->SetAuger(true);
+  atom_->SetCutForSecondaryPhotons(0);
```
Here the class is initialized with two class members (`minGammaEnergy` and `minElectronEnergy`) to zero. However, when you generate the particles in line 149, you comment that 0 can't be used and those variable are modified inside the function. Is there any reason why you should not set `0.1 eV` here?

gonzaponte replied:
I think this was just the first thing I tried and then I realized that I needed to set it to >0 in the GenerateParticles function. At the time I didn't know that GenerateParticles overwrites these values. So probably these two lines can even be removed. What do you prefer?

paolafer replied:
For what I can see in the code, those thresholds are used exclusively in the `GenerateParticles()` function, therefore it should be exactly the same; on top of that, that function requires to be passed those arguments. The whole structure looks quite confusing to me, but it's Geant4... I would remove the setters at the beginning, since they don't do anything.

gonzaponte replied:
As discussed offline, these setters have no effect as the `GenerateParticle` method overrides them. They have been removed.

gonzaponte committed:
Remove calls to cut setters in initialization

These values are overriden by the arguments used in the function
`GenerateParticles`

paolafer reviewed (approved):
This PR adds the code for the double electron capture generator, used in the simulations of the NEXT paper on ECEC. It was long due and it's finally approved!
